independent-projects/qute/core/src/main/java/io/quarkus/qute/TemplateInstance.java

```diff
@@ -37,7 +37,9 @@ public interface TemplateInstance {
      * @param data
      * @return
      */
-    TemplateInstance data(Object data);
+    default TemplateInstance data(Object data) {
+        throw new UnsupportedOperationException();
```

Comment:
All these default implementations are a bit weird but we need this in order to be able to use the interface for template records (so that users don't need to implement the methods themselves). 

Alternatively, we would need to introduce a new marker interface that would extend `TemplateInstance` and define default implementations for all these methods.

NOTE: Records may not extend classes but only implement interfaces.

Reply:
Not sure I like requiring records to implement TemplateInstance. Perhaps it's easier if we had a static method in `TemplateInstance` wrapping records (eg. `TemplateInstance.forRecord(new item(...))`?

Reply:
OTOH it's simpler than the inner class + annotation strategy, so maybe it's okay

Reply:
Can't we just assume records as template instances intrinsically?

Reply:
> Can't we just assume records as template instances intrinsically?

That wouldn't help in cases where you want to use the `TemplateInstance` methods; such as `new HelloWolrd(name).render()`...

Also we use the `TemplateInstance` as a marker, so we would have to make the `@CheckedTemplate` mandatory which is IMO :-1: (right now, the  `@CheckedTemplate` for records is only needed for non-default config).

Reply:
> Not sure I like requiring records to implement TemplateInstance.

BTW we require the `static native` methods of a `@CheckedTemplate` to return `TemplateInstance`. So it's a very similar requirement IMO.

Reply:
Damn, so I guess I majorly fucked up by never pushing my branch which had support for this. Though I'm sure you did it better :)

I did wonder about these default interface methods, and came to a slightly different solution of declaring a new subinterface of `TemplateInstance` called `View` (bad name, but it was shorter). That interface overrode every `TemplateInstance` method to make it `default` and forward to a single method `wrapped` that the records had to implement.

```
public interface View extends TemplateInstance {

    public default TemplateInstance wrapped() {
        throw new RuntimeException("This should be generated");
    }

    @Override
    public default TemplateInstance data(Object data) {
        return wrapped().data(data);
    }

    @Override
    public default TemplateInstance data(String key, Object data) {
        return wrapped().data(key, data);
    }

    @Override
    public default TemplateInstance setAttribute(String key, Object value) {
        return wrapped().setAttribute(key, value);
    }

    @Override
    public default Object getAttribute(String key) {
        return wrapped().getAttribute(key);
    }

    @Override
    public default String render() {
        return wrapped().render();
    }

    @Override
    public default CompletionStage<String> renderAsync() {
        return wrapped().renderAsync();
    }

    @Override
    public default Multi<String> createMulti() {
        return wrapped().createMulti();
    }

    @Override
    public default Uni<String> createUni() {
        return wrapped().createUni();
    }

    @Override
    public default CompletionStage<Void> consume(Consumer<String> consumer) {
        return wrapped().consume(consumer);
    }

    @Override
    public default long getTimeout() {
        return wrapped().getTimeout();
    }

    @Override
    public default Template getTemplate() {
        return wrapped().getTemplate();
    }

    @Override
    public default TemplateInstance onRendered(Runnable action) {
        return wrapped().onRendered(action);
    }
}
```

And records were desugared to:

```java
class Controller {
    static class foo implements View {

        private final String str;
        private final int i;
        // added
        private final TemplateInstance wrapped;

        public foo(String str, int i) {
            this.str = str;
            this.i = i;
            Template template = Arc.container().instance(TemplateProducer.class).get().getInjectableTemplate("Controller/foo");
            TemplateInstance instance = template.instance();
            instance = instance.data("str", str);
            instance = instance.data("i", i);
            this.wrapped = instance;
        }

        @Override
        public TemplateInstance wrapped() {
            return wrapped;
        }
    }
}
```

Looks a lot like the bytecode you produce, except you override every method to forward the calls.

Now, my implementation falls down when it comes to supporting `MailTemplateInstance` unless I introduce another interface called `MailView` which implements both `MailTemplateInstance` and `View`, I guess.

Anyway, thought I'd put it out there for you to look at and consider. I don't think my strategy was better. I think yours is more logical (I was afraid you'd object to making all methods of `TemplateInstance` `default`, hence the new interface, but you did it anyway :) ). Unless something strikes you as interesting, we can throw the idea away.